File: computing/interview/design-patterns/behavioral/template-method/java/TemplateMethodDemo.java
package interview.designpatterns.behavioral.templatemethod;

import java.util.Locale;

abstract class DataProcessor {
    public final void process(String input) {
        String sanitized = sanitize(input);
        String transformed = transform(sanitized);
        persist(transformed);
    }

    protected String sanitize(String input) {
        return input.replaceAll("\\s", "");
    }

    protected abstract String transform(String sanitized);
    protected abstract void persist(String transformed);
}

final class UppercaseProcessor extends DataProcessor {
    @Override
    protected String transform(String sanitized) {
        return sanitized.toUpperCase(Locale.ROOT);
    }

    @Override
    protected void persist(String transformed) {
        System.out.printf("Persisting uppercase string: %s%n", transformed);
    }
}

final class HashProcessor extends DataProcessor {
    @Override
    protected String transform(String sanitized) {
        int hash = 17;
        for (char c : sanitized.toCharArray()) {
            hash = hash * 31 + c;
        }
        return Integer.toString(hash);
    }

    @Override
    protected void persist(String transformed) {
        System.out.printf("Persisting hash: %s%n", transformed);
    }
}

public final class TemplateMethodDemo {
    private TemplateMethodDemo() {}

    public static void main(String[] args) {
        DataProcessor upper = new UppercaseProcessor();
        DataProcessor hash = new HashProcessor();

        upper.process("  Hello Template Method  ");
        hash.process("  Hello Template Method  ");
    }
}

File: computing/interview/design-patterns/behavioral/template-method/csharp/TemplateMethodDemo.cs
using System;
using System.Linq;

namespace Interview.DesignPatterns.Behavioral.TemplateMethod
{
    public abstract class DataProcessor
    {
        public void Process(string input)
        {
            var cleaned = Sanitize(input);
            var transformed = Transform(cleaned);
            Persist(transformed);
        }

        protected virtual string Sanitize(string input)
        {
            return new string(input.Where(c => !char.IsWhiteSpace(c)).ToArray());
        }

        protected abstract string Transform(string sanitized);
        protected abstract void Persist(string transformed);
    }

    public sealed class UppercaseProcessor : DataProcessor
    {
        protected override string Transform(string sanitized) => sanitized.ToUpperInvariant();

        protected override void Persist(string transformed)
        {
            Console.WriteLine($"Persisting uppercase string: {transformed}");
        }
    }

    public sealed class HashProcessor : DataProcessor
    {
        protected override string Transform(string sanitized)
        {
            return sanitized.Aggregate(17, (hash, c) => hash * 31 + c).ToString();
        }

        protected override void Persist(string transformed)
        {
            Console.WriteLine($"Persisting hash: {transformed}");
        }
    }

    public static class Demo
    {
        public static void Main()
        {
            DataProcessor upper = new UppercaseProcessor();
            DataProcessor hash = new HashProcessor();

            upper.Process("  Hello Template Method  ");
            hash.Process("  Hello Template Method  ");
        }
    }
}
